### PR TITLE
dynamically gather certs as a side-car

### PR DESCRIPTION
**Pull request by deads2k**
This updates the static pod to have a side-car container that is synchronizing configmaps to a target location.

@sttts @mfojtik

### CODE REVIEW

File: bindata/v3.11.0/kube-apiserver/kubeconfig-cm.yaml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+kind: ConfigMap
+metadata:
+  name: kube-apiserver-cert-syncer-kubeconfig
+  namespace: openshift-kube-apiserver
+data:
+  kubeconfig: |
+    apiVersion: v1
+    clusters:
+      - cluster:
+          certificate-authority: /etc/kubernetes/static-pod-resources/configmaps/kube-apiserver-server-ca/ca-bundle.crt
```
Comment (mfojtik) [inline bindata/v3.11.0/kube-apiserver/kubeconfig-cm.yaml]: @deads2k shouldn't this be "certs" ?

File: bindata/v3.11.0/kube-apiserver/kubeconfig-cm.yaml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+          cluster: loopback
+          user: kube-apiserver-cert-syncer
+        name: kube-apiserver-cert-syncer
+    current-context: kube-apiserver-cert-syncer
+    kind: Config
+    preferences: {}
+    users:
+      - name: kube-apiserver-cert-syncer
+        user:
+          client-certificate: /etc/kubernetes/static-pod-resources/secrets/kube-apiserver-cert-syncer-client-cert-key/tls.crt
```
Comment (mfojtik) [inline bindata/v3.11.0/kube-apiserver/kubeconfig-cm.yaml]: same here?

File: bindata/v3.11.0/kube-apiserver/pod.yaml [hunk 3]
```diff
@@ -50,6 +52,33 @@ spec:
...
+      valueFrom:
+        fieldRef:
+          fieldPath: metadata.namespace
+    image: ${OPERATOR_IMAGE}
+    imagePullPolicy: IfNotPresent
+    terminationMessagePolicy: FallbackToLogsOnError
+    command: ["cluster-kube-apiserver-operator", "cert-syncer"]
+    args:
+      - --kubeconfig=/etc/kubernetes/static-pod-resources/configmaps/kube-apiserver-cert-syncer-kubeconfig/kubeconfig
+      - --namespace=${POD_NAMESPACE}
```
Comment (mfojtik) [inline bindata/v3.11.0/kube-apiserver/pod.yaml]: nit: i think you can read the namespace from inside pod (/var/run/...)

File: cmd/cluster-kube-apiserver-operator/main.go [hunk 4]
```diff
@@ -7,18 +7,18 @@ import (
...
-	"github.com/openshift/cluster-kube-apiserver-operator/pkg/cmd/operator"
+	"github.com/openshift/cluster-kube-apiserver-operator/pkg/cmd/certsync"
+	operatorcmd "github.com/openshift/cluster-kube-apiserver-operator/pkg/cmd/operator"
 	"github.com/openshift/cluster-kube-apiserver-operator/pkg/cmd/render"
 	"github.com/openshift/cluster-kube-apiserver-operator/pkg/cmd/resourcegraph"
+	"github.com/openshift/cluster-kube-apiserver-operator/pkg/operator"
 	"github.com/openshift/cluster-kube-apiserver-operator/pkg/version"
 	"github.com/openshift/library-go/pkg/operator/staticpod/installerpod"
 	"github.com/openshift/library-go/pkg/operator/staticpod/prune"
+	"github.com/spf13/cobra"
```
Comment (mfojtik) [inline cmd/cluster-kube-apiserver-operator/main.go]: nit: move these up

File: pkg/cmd/certsync/certsync_cmd.go [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"reflect"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/openshift/library-go/pkg/config/client"
```
Comment (mfojtik) [inline pkg/cmd/certsync/certsync_cmd.go]: nit: order

File: pkg/cmd/certsync/certsync_cmd.go [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+			if err != nil {
+				glog.Fatal(err)
+			}
+			r.Run(1, make(chan struct{}))
+		},
+	}
+
+	cmd.Flags().StringVar(&o.DestinationDir, "destination-dir", o.DestinationDir, "Directory to write to")
+	cmd.Flags().StringVarP(&o.Namespace, "namespace", "n", o.Namespace, "Namespace to read from")
+	cmd.Flags().StringVar(&o.KubeConfigFile, "kubeconfig", o.KubeConfigFile, "Location of the master configuration file to run from.")
```
Comment (mfojtik) [inline pkg/cmd/certsync/certsync_cmd.go]: master configuration file?

File: pkg/cmd/certsync/certsync_cmd.go [hunk 7]
```diff
@@ -0,0 +1,272 @@
...
+func NewSyncController(configmaps, secrets []revision.RevisionResource) *cobra.Command {
+	o := &CertSyncControllerOptions{
+		configMaps: configmaps,
+		secrets:    secrets,
+	}
+
+	cmd := &cobra.Command{
+		Use: "cert-syncer --kubeconfig=kubeconfigfile",
+		Run: func(cmd *cobra.Command, args []string) {
+			r, err := o.Complete()
```
Comment (mfojtik) [inline pkg/cmd/certsync/certsync_cmd.go]: nit: missing Validate which makes @soltysh sad

File: pkg/cmd/certsync/certsync_cmd.go [hunk 8]
```diff
@@ -0,0 +1,272 @@
...
+
+	return cmd
+}
+
+func (o *CertSyncControllerOptions) Complete() (*CertSyncController, error) {
+	kubeConfig, err := client.GetKubeConfigOrInClusterConfig(o.KubeConfigFile, nil)
+	if err != nil {
+		return nil, err
+	}
+	protoKubeConfig := rest.CopyConfig(kubeConfig)
```
Comment (mfojtik) [inline pkg/cmd/certsync/certsync_cmd.go]: reminds me we need some library-go helper for this

File: bindata/v3.11.0/kube-apiserver/kubeconfig-cm.yaml (hunk 1 above)
Comment (deads2k) [reply bindata/v3.11.0/kube-apiserver/kubeconfig-cm.yaml]: > @deads2k shouldn't this be "certs" ?

we dont' dynamically reload this yet

File: pkg/cmd/certsync/certsync_cmd.go [hunk 9]
```diff
@@ -0,0 +1,272 @@
...
+	protoKubeConfig.ContentType = "application/vnd.kubernetes.protobuf"
+
+	// This kube client use protobuf, do not use it for CR
+	kubeClient, err := kubernetes.NewForConfig(protoKubeConfig)
+	if err != nil {
+		return nil, err
+	}
+	kubeInformers := informers.NewSharedInformerFactoryWithOptions(kubeClient, 10*time.Minute, informers.WithNamespace(o.Namespace))
+
+	eventRecorder := events.NewKubeRecorder(kubeClient.CoreV1().Events(o.Namespace), "cert-syncer",
```
Comment (mfojtik) [inline pkg/cmd/certsync/certsync_cmd.go]: why not use: 
```
eventTarget, err = events.GetControllerReferenceForCurrentPod(o.KubeClient, o.Namespace, nil)
recorder := events.NewRecorder(o.KubeClient.CoreV1().Events(o.Namespace), "cert-syncer", eventTarget)
```

you already set the env vars for this to work out of box

File: pkg/cmd/certsync/certsync_cmd.go [hunk 10]
```diff
@@ -0,0 +1,272 @@
...
+			}
+			continue
+		case err != nil:
+			errors = append(errors, err)
+			continue
+		}
+
+		contentDir := getSecretDir(c.destinationDir, s.Name)
+		glog.Infof("Creating directory %q ...", contentDir)
+		if err := os.MkdirAll(contentDir, 0755); err != nil && !os.IsExist(err) {
```
Comment (mfojtik) [inline pkg/cmd/certsync/certsync_cmd.go]: isn't 755 generous for secret content?

File: pkg/operator/targetconfigcontroller/targetconfigcontroller.go [hunk 11]
```diff
@@ -225,6 +228,9 @@ func managePod(client coreclientv1.ConfigMapsGetter, recorder events.Recorder, o
 			required.Spec.InitContainers[0].Image = imagePullSpec
 		}
 	}
+	if len(operatorImagePullSpec) > 0 {
```
Comment (mfojtik) [inline pkg/operator/targetconfigcontroller/targetconfigcontroller.go]: do we ever have case when this is not set?

File: pkg/cmd/certsync/certsync_cmd.go (hunk 9 above)
Comment (deads2k) [reply pkg/cmd/certsync/certsync_cmd.go]: > you already set the env vars for this to work out of box

not attached to an RS, it panics

File: pkg/cmd/certsync/certsync_cmd.go (hunk 10 above)
Comment (deads2k) [reply pkg/cmd/certsync/certsync_cmd.go]: > isn't 755 generous for secret content?

There's an existing todo

File: pkg/operator/targetconfigcontroller/targetconfigcontroller.go (hunk 11 above)
Comment (deads2k) [reply pkg/operator/targetconfigcontroller/targetconfigcontroller.go]: > do we ever have case when this is not set?

improper wiring

**Comment by deads2k** [comment]
This is green.  I plan move the library-go code up, bump, merge, and followup anything else.

**Comment by deads2k** [comment]
all green, nothing merged

**Pull request merged**